### PR TITLE
Adjust blue warp item behavior

### LINKED ISSUE #2230: Blue warps sometimes don't activate
The new blue warps from #2034 sometimes don't activate. Only 3 reports since this was added to my fork in January 2023, most recently at <https://www.twitch.tv/videos/2157998297> played on 8.1.39. When this happens, reloading the scene fixes the bug, but also duplicates the item given by the blue warp.

A previous idea of what could be causing this was that the ranges for giving the item and for activating the blue warp might not exactly match and that this is what causes the bug (https://github.com/OoTRandomizer/OoT-Randomizer/pull/2034#issuecomment-1627584163), but I was unable to reproduce by drastically changing one radius or the other.

Given the camera angle in this most recent clip, maybe it has something to do with the center of the blue warp not being on screen at the time the item is picked up?

### PR DESCRIPTION
**Pull request by fenhl**
This PR changes how dungeon rewards give out their items: A “rewards as items” flag is introduced, which works similarly to the existing “songs as items” and “fairy ocarina as item” flags.

* When disabled, blue warps now simply run the item's effect function and display its text box. This has the nice side effect of removing the interruption from the get item animation while keeping the speedups introduced in #2034.
* When enabled, the item is now added to the delayed item queue, causing it to be given after warping.

Neither behavior will prevent the blue warp entrance from activating. If my analysis at https://github.com/OoTRandomizer/OoT-Randomizer/issues/2230#issuecomment-2141059686 is correct, this fixes #2230.

### CODE REVIEW

**Comment by fenhl** [comment]
Here's what it looks like without rewards as items:

https://github.com/OoTRandomizer/OoT-Randomizer/assets/641386/7349ef8d-bc5a-4544-accd-4a8624f42ac5

**Comment by fenhl** [comment]
Can confirm this fixes the bug with the setup from https://github.com/OoTRandomizer/OoT-Randomizer/issues/2230#issuecomment-2193696608 without rewards as items:

[OoT A7A63 FMWRA80NZT.webm](https://github.com/OoTRandomizer/OoT-Randomizer/assets/641386/d445448b-8db2-4b47-9001-3807978fec44)

Howewer, the rewards as items flag doesn't seem to be set correctly:

[OoT 5357C 60ZC6AXQ6Y.webm](https://github.com/OoTRandomizer/OoT-Randomizer/assets/641386/1dc30110-a131-4386-ab43-844a5b0b2629)

**Comment by fenhl** [comment]
> Howewer, the rewards as items flag doesn't seem to be set correctly:

This turned out to be a bug with the size of `CFG_BIGOCTO_OVERRIDE_KEY` not having been adjusted for #2069. Now that that's fixed, rewards as items works correctly as well:

[OoT 5357C PQ2VHJI0DP.webm](https://github.com/OoTRandomizer/OoT-Randomizer/assets/641386/4dc26ce6-79e1-4155-95b0-305073fe3b7d)

File: ASM/c/blue_warp.c [hunk 1]
```diff
@@ -18,38 +23,28 @@ int32_t DoorWarp1_PlayerInRange(z64_actor_t* actor, z64_game_t* game) {
...
+            if (REWARDS_AS_ITEMS) {
+                push_delayed_item(0x05 + boss_idx);
+            } else {
+                override_key_t override_key = { .scene = 0xFF, .type = OVR_DELAYED, .flag = 0x05 + boss_idx };
+                override_t override = lookup_override_by_key(override_key);
+                uint16_t resolved_item_id = resolve_upgrades(override);
+                item_row_t* item_row = get_item_row(resolved_item_id);
+                call_effect_function(item_row);
+                PLAYER_NAME_ID = override.value.base.player;
+                z64_DisplayTextbox(&z64_game, resolve_item_text_id(item_row, PLAYER_NAME_ID != PLAYER_ID), 0);
```
Comment (flagrama) [inline ASM/c/blue_warp.c:41]: Does this basically amount to "display a non-blocking textbox while the warp is activating"?

File: ASM/c/blue_warp.c [hunk 2]
```diff
@@ -18,38 +23,28 @@ int32_t DoorWarp1_PlayerInRange(z64_actor_t* actor, z64_game_t* game) {
...
             }
-            return true;
+            extended_savectx.collected_dungeon_rewards[boss_idx] = true;
```
Comment (flagrama) [inline ASM/c/blue_warp.c:43]: This isn't susceptible to someone pausing, saving, resetting frame perfectly after the warp is it? The delayed item stuff is stored in the (extended) save context now so even after a reset they'll receive the item?

File: LocationList.py [hunk 3]
```diff
@@ -62,15 +62,15 @@ def shop_address(shop_id: int, shelf_id: int) -> int:
...
-    ("Morpha",                                                       ("Boss",         0x16,  0x65, None,                            'Water Medallion',                       ("Water Temple", "Water Temple MQ", "Vanilla Dungeons", "Master Quest", "Dungeon Rewards",))),
-    ("Bongo Bongo",                                                  ("Boss",         0x18,  0x65, None,                            'Shadow Medallion',                      ("Shadow Temple", "Shadow Temple MQ", "Vanilla Dungeons", "Master Quest", "Dungeon Rewards",))),
-    ("Twinrova",                                                     ("Boss",         0x17,  0x65, None,                            'Spirit Medallion',                      ("Spirit Temple", "Spirit Temple MQ", "Vanilla Dungeons", "Master Quest", "Dungeon Rewards",))),
+    ("ToT Reward from Rauru",                                        ("Boss",         0xFF,  0x04, None,                            'Light Medallion',                       ("Temple of Time", "NPCs", "Dungeon Rewards",))),
```
Comment (flagrama) [inline LocationList.py:67]: I know this is to simplify the logic in several places, but it seems weird to have it be the "Boss" type. Perhaps it's time to rename this type to something else? Probably not in this PR, but maybe an issue for doing it in the future?

**Review by flagrama — COMMENTED** [review]
Just some clarifying questions really. Otherwise the code LGTM.

File: ASM/c/blue_warp.c (hunk 1 above)
Comment (fenhl) [reply ASM/c/blue_warp.c:41]: That and run the item's effect function, which adds the dungeon reward to the quest status.

File: ASM/c/blue_warp.c (hunk 2 above)
Comment (fenhl) [reply ASM/c/blue_warp.c:43]: Correct.

File: LocationList.py (hunk 3 above)
Comment (fenhl) [reply LocationList.py:67]: It used to be categorized like this but was changed in #2034. That change is no longer necessary now.

**Review by cjohnson57 — APPROVED** [review]

**Pull request merged**